### DISCUSSION TITLE
[Feature request] Better Integration with Existing Templates

### DISCUSSION THREAD

**Discussion by motion2082** [danielo515/obsidian-modal-form#350] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

- I think this Plugin has a lot of potential. 
- The issue is how it integrates with Templater and QuickAdd
- Making a Form and applying it to a new note is easy, applying it to an existing Template (Templater or QuickAdd) is difficult and clunky
- If I have a template that needs Templater script executed, loading a Modal form to update the fields is really difficult

**Describe the solution you'd like**
- Better integration with existing Templates 
- Easy Update Variables in existing Template option

**Comment by danielo515** [comment]
Can you give me an example where you are finding difficulties integrating a form into an existing template?

**Comment by motion2082** [comment]
Here is one example of a Templater Template that's already been created in Vault. Hard to update via Modal Form. You can use it in Template Builder but tR += result.asFrontmatterString(); doesn't work

```
<%*
  const dateAdded = tp.date.now('YYYY-MM-DD');
const week = tp.date.now("gggg-[W]ww");
const quarter = `Q${tp.date.now("Q")} - ${tp.date.now("YYYY")}`;
const month = tp.date.now("MMMM YYYY");
const year = tp.date.now("YYYY");

const metaData = `---
tags:
  - "#My/Achievement"
  - "#🌟"
cssclasses:
aliases:
Date-Added: "[[${dateAdded}]]"
Week: "[[${week}]]"
Quarter: "[[${quarter}]]"
Month: "[[${month}]]"
Year: "[[${year}]]"
Summary: 
What-Worked: 
---`;
-%>

>[!info]- Meta Details
> Date: [[<% tp.date.now("DD-MM-YY") %>]]
> Week: [[<% tp.date.now("gggg-[W]ww") %>]]
> Month: [[<% tp.date.now("MMMM YYYY") %>]]
> Quarter: [[Q<% tp.date.now("Q") %> - <% tp.date.now("YYYY") %>]]
> Year: [[<% tp.date.now("YYYY") %>]]
```

Trying to update with Modal Form to Add the YAML (summary and what-worked)

Summary: 
What-Worked: 

Also if I had a heading could you also update below it with Modal Form?

## Achievement Summary

Add Summary with Modal Form here

**Comment by motion2082** [comment]
So I figured out one potential way to do it 

Leave the YAML Properties (Source/Summary) out of the original Template and Add them with a QuickAdd Capture in a Macro after the Template is created

` Capture to add `
` Summary: `
` Source: `

```js quickadd
    const modalForm = app.plugins.plugins.modalforms.api;
    const result = await modalForm.openForm('Test');
    return result.asFrontmatterString();
```

Just means that YAML isn't seen in original template.

Also wouldn't be able to update the Data with Modal Form after the Macro is run 

This is a work around though

**Comment by motion2082** [comment]
Playing around with some Templater Code, it's making a little more sense with trial and error.

**Comment by danielo515** [comment]
Taking your template example. You want to create a note using that template, and later update the note using modal forms?

Or you want to insert the template when creating the note and, at the same time show the modal to collect those additional details and then get the complete note in one go? If that is the case, then it is very simple and I will try to provide an example later when I arrive at home

**Comment by motion2082** [comment]
Hi Danielo,

i did manage to get it working with the existing templater template.

Is there a way to update the note with Modal Form once the note has been created?

So YAML and note variables set with Modal then run again to update?

**Comment by danielo515** [comment, marked as answer]
Does your note look something like this now?

```js
<%*
  const dateAdded = tp.date.now('YYYY-MM-DD');
const week = tp.date.now("gggg-[W]ww");
const quarter = `Q${tp.date.now("Q")} - ${tp.date.now("YYYY")}`;
const month = tp.date.now("MMMM YYYY");
const year = tp.date.now("YYYY");

const modalForm = app.plugins.plugins.modalforms.api;
const result = await modalForm.openForm('Test');

const metaData = `---
tags:
  - "#My/Achievement"
  - "#🌟"
cssclasses:
aliases:
Date-Added: "[[${dateAdded}]]"
Week: "[[${week}]]"
Quarter: "[[${quarter}]]"
Month: "[[${month}]]"
Year: "[[${year}]]"
${result.asFrontmatterString()}
---`;
-%>

>[!info]- Meta Details
> Date: [[<% tp.date.now("DD-MM-YY") %>]]
> Week: [[<% tp.date.now("gggg-[W]ww") %>]]
> Month: [[<% tp.date.now("MMMM YYYY") %>]]
> Quarter: [[Q<% tp.date.now("Q") %> - <% tp.date.now("YYYY") %>]]
> Year: [[<% tp.date.now("YYYY") %>]]
```

That will create the note, open the form, and finally populate the details you want in the frontmatter.



> Is there a way to update the note with Modal Form once the note has been created?
> 
> So YAML and note variables set with Modal then run again to update?

It is possible to update the YAML part, because it is easily machine readable, but not the rest of the note. For doing that you will need a template dedicated to edit the YAML part of a note. Note that templates can be used to trigger commands, they don't always need to insert any data in the current note. Here is an example:

https://danielorodriguez.com/obsidian-modal-form/advanced-examples/#modifying-frontmatter-with-a-form

Let me know if you need more help

**Comment by motion2082** [comment]
Sort of :)

Instead of ${result.asFrontmatterString()} after Year:

I used QuickAdd Capture Choice with

```js quickadd
    const modalForm = app.plugins.plugins.modalforms.api;
    const result = await modalForm.openForm('Achievement');
    return result.asFrontmatterString();
```

after Year:

Only because I was calling Template with QuickAdd first as I don't use Templater to call/create templates

**Comment by motion2082** [comment]
I just tested your method as a Template Choice only in QuickAdd (no Capture) and that also works

Legend thank you!

**Reply by danielo515** [reply]
Happy to see that works!

**Comment by Dawe5ome** [comment]
Hi so I am trying to have the modal forms plugin run when I call it via a metabind button on my daily note. I was successfully able to get this to work and have the form fields in my YAML, however, what I am struggling is to get the certain form fields to populate not just in my YAML but also in my the body of my note under the field. 

To simplify what I want my process for daily notes to look like is to:
1. Create a Daily note at the beginning of my day to access my tasks and habit trackers
2. Hit the metabind button to call my journal form at the end of the day and have it populate my front matter and in my # Journal section

Here is my Daily Note Template
```
---
tags:
  - dailynote
date: <% tp.date.now() %>
week: <% tp.date.now("[W]ww") %>
year: <% tp.date.now("YYYY") %>
aka: 
enjoyment: 
mood: 
today: 
nourished: 
mastered: 
depleted:
---
<% tp.web.daily_quote() %>
[[<% tp.date.now("YYYY-MM-DD",-1) %>|⬅️Yesterday]] - [[<% tp.date.now("YYYY-MM-DD",1) %>|Tomorrow➡️]]
[[<% tp.date.now("YYYY-[W]ww") %>|This week]] 

`BUTTON[DailyNote]`
# Journal
## What did I do today?
<% result.get("today") %>

## What made me feel *Nourished?*
<% result.get("nourished") %>


## What made me feel *Mastered?*
<% result.get("mastered") %>


## What made me feel *Depleted?*
<% result.get("depleted") %>
```

Here is the template for the form that I can call to
```
<%*
const modalForm = app.plugins.plugins.modalforms.api;

const run = async (frontmatter) => {
  // Open your "DailyNote" form, preloading it with the current frontmatter values
  const result = await modalForm.openForm('DailyNote', {
    values: { ...frontmatter },
  });
  return result.getData(); // Return the updated form data
};

// Get the updated data from the form
const data = await run(tp.frontmatter);

// Update the frontmatter with the new data
app.fileManager.processFrontMatter(
  tp.config.target_file,
  frontmatter => {
    Object.assign(frontmatter, data);
  },
);
%>
```

**Reply by danielo515** [reply]
@Dawe5ome that looks like a separate question from this original one. Do you mind opening a new discussion? You  don't need to specify anything, just copy paste the content of this one, so it will be easier to find in the future for people struggling with the same thing.

**Reply by danielo515** [reply]
I will answer you anyway, but please open the new discussion and ping me there so I can also copy paste my answer.
Let me state some facts before:
1. The easiest path to everything is to create each note in one go. This means, a single template generates the entire note in one single and atomic operation. I understand this not always fits in every workflow, but it's worth considering if you can change your task a bit to fulfil this simpler scenario.
2. It is usually better to have a single source of truth, specially if you plan to do changes to the data. Having the same data copied in several places means it can get out of sync. This little fact is the key of one of the solutions I'm gonna suggest.
3. Last but not least, templates can both run code and output content, and we can take advantage of that.

So, first solution I'm proposing is to just do everything at the end of the day, create the file at the same time you fill the form in a single template and do it in a single operation. If this does not fit your use case, go to the next.

Another solution would be to use the frontmatter/yaml as the source of the information, and just reference it using dataview (this is an awesome feature of  DV). So your template could looks something like this:

```md
---
tags:
  - dailynote
date: <% tp.date.now() %>
week: <% tp.date.now("[W]ww") %>
year: <% tp.date.now("YYYY") %>
aka: 
enjoyment: 
mood: 
today: 
nourished: 
mastered: 
depleted:
---
<% tp.web.daily_quote() %>
[[<% tp.date.now("YYYY-MM-DD",-1) %>|⬅️Yesterday]] - [[<% tp.date.now("YYYY-MM-DD",1) %>|Tomorrow➡️]]
[[<% tp.date.now("YYYY-[W]ww") %>|This week]] 

`BUTTON[DailyNote]`
# Journal
## What did I do today?
`= this.today`

## What made me feel *Nourished?*
`= this.nourished`


## What made me feel *Mastered?*
`= this.mastered`


## What made me feel *Depleted?*
`= this.depleted`
```

You can check the docs of [this feature here](https://blacksmithgu.github.io/obsidian-dataview/querieske/dql-js-inline/#inline-dql).
That will show nothing at first, but should show the contents of those fields on the yaml.

The last solution I propose is to "embed" the entire contents of your note's content into the modal form template, so it gets inserted when you fill your form, and not beforehand like you are doing now. This will be your templates for this solution:

```md
---
tags:
  - dailynote
date: <% tp.date.now() %>
week: <% tp.date.now("[W]ww") %>
year: <% tp.date.now("YYYY") %>
aka: 
enjoyment: 
mood: 
today: 
nourished: 
mastered: 
depleted:
---
<% tp.web.daily_quote() %>
[[<% tp.date.now("YYYY-MM-DD",-1) %>|⬅️Yesterday]] - [[<% tp.date.now("YYYY-MM-DD",1) %>|Tomorrow➡️]]
[[<% tp.date.now("YYYY-[W]ww") %>|This week]] 

`BUTTON[DailyNote]`
```

```js
<%*
const modalForm = app.plugins.plugins.modalforms.api;

const run = async (frontmatter) => {
  // Open your "DailyNote" form, preloading it with the current frontmatter values
  const result = await modalForm.openForm('DailyNote', {
    values: { ...frontmatter },
  });
  return result.getData(); // Return the updated form data
};

// Get the updated data from the form
const data = await run(tp.frontmatter);

// Update the frontmatter with the new data
app.fileManager.processFrontMatter(
  tp.config.target_file,
  frontmatter => {
    Object.assign(frontmatter, data);
  },
);
%>

# Journal
## What did I do today?
<% result.get("today") %>

## What made me feel *Nourished?*
<% result.get("nourished") %>


## What made me feel *Mastered?*
<% result.get("mastered") %>


## What made me feel *Depleted?*
<% result.get("depleted") %>
```

Let me know if any of those solutions works for you.

**Reply by Dawe5ome** [reply]
@danielo515 Thank you for your quick response and answer! After giving all your solutions a try I found that the best solution that works best with my workflow was the second solution which was to use data view. I did have issues accessing your document "You can check the docs of [this feature here](https://blacksmithgu.github.io/obsidian-dataview/querieske/dql-js-inline/#inline-dql)." however it did provide the best solution for my flow without interrupting what I was doing. 

I really did like the third iteration of the solution however, I found that I could simply not get it to work with the second template. Thank you again for your quick response :)

**Reply by danielo515** [reply]
>  could simply not get it to work with the second template.

What do you mean with that? 
The third solution uses two templates, the first one is the daily note template, which should be created on a daily basis or whenever you click a button. 
Then, once that first step completed and saved to your note, need to insert inline the second template. You can try this in the command palette, insert templater template and select that template. That should work. Maybe the problem is that metabind button is not inserting the template, just executing it?
